File: java/array_search/JumpSearchProduction.java
import java.util.Comparator;
import java.util.List;

public final class JumpSearchProduction {
    public record SearchResult(boolean found, int index, int insertionPoint, int blockSize) {
    }

    private JumpSearchProduction() {
    }

    public static <T extends Comparable<? super T>> SearchResult jumpSearchFirst(List<T> values, T target) {
        return jumpSearchFirst(values, target, Comparator.naturalOrder(), 0);
    }

    public static <T extends Comparable<? super T>> SearchResult jumpSearchFirst(List<T> values, T target, int blockSize) {
        return jumpSearchFirst(values, target, Comparator.naturalOrder(), blockSize);
    }

    public static <T> SearchResult jumpSearchFirst(List<T> values, T target, Comparator<? super T> comparator, int blockSize) {
        if (values == null || comparator == null) {
            throw new IllegalArgumentException("values and comparator must not be null");
        }

        int length = values.size();

        if (length == 0) {
            return new SearchResult(false, -1, 0, 0);
        }

        int step = blockSize > 0 ? blockSize : (int) Math.sqrt(length);

        if (step < 1) {
            step = 1;
        }

        int left = 0;
        int right = Math.min(step, length);

        while (right < length && comparator.compare(values.get(right - 1), target) < 0) {
            left = right;
            right = Math.min(right + step, length);
        }

        int insertionPoint = lowerBound(values, target, left, right, comparator);
        boolean found = insertionPoint < length && comparator.compare(values.get(insertionPoint), target) == 0;

        return new SearchResult(found, found ? insertionPoint : -1, insertionPoint, step);
    }

    private static <T> int lowerBound(List<T> values, T target, int left, int right, Comparator<? super T> comparator) {
        while (left < right) {
            int mid = left + (right - left) / 2;

            if (comparator.compare(values.get(mid), target) < 0) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return left;
    }
}

File: python/array_search/jump_search_production.py
from dataclasses import dataclass 
from math import isqrt


@dataclass(frozen=True)
class JumpSearchResult: 
    found: bool 
    index: int 
    insertion_point: int 
    block_size: int 


def _lower_bound(values, target, left, right, key):
    while left < right: 
        mid = left + (right - left) // 2

        if key(values[mid]) < target:
            left = mid + 1
        else:
            right = mid 

    return left 


def jump_search_first(values, target, *, key=None, block_size=None):
    key_fn = key or (lambda value: value)
    length = len(values)

    if length == 0: 
        return JumpSearchResult(False, -1, 0, 0) 

    step = block_size or isqrt(length)

    if step < 1: 
        raise ValueError("block_size must be positive")

    left = 0
    right = min(step, length)

    while right < length and key_fn(values[right - 1]) < target:
        left = right 
        right = min(right + step, length)

    insertion_point = _lower_bound(values, target, left, right, key_fn)
    found = insertion_point < length and key_fn(values[insertion_point]) == target

    return JumpSearchResult(
        found,
        insertion_point if found else -1,
        insertion_point,
        step
    )
